sympy/stats/stochastic_process_types.py

```diff
@@ -224,12 +234,12 @@ def joint_distribution(self, *args):
                 raise ValueError("Expected a RandomIndexedSymbol or "
                                 "key not  %s"%(type(arg)))
 
-        if args[0].pspace.distribution == None: # checks if there is any distribution available
-            return JointDistribution(*args)
-
-        pdf = Lambda(tuple(args),
-                expr=Mul.fromiter(arg.pspace.process.density(arg) for arg in args))
-        return JointDistributionHandmade(pdf)
+        proc = args[0].pspace.process
+        if hasattr(proc, 'density'):
```

Comment:
Instead, define a `density` method in `StochasticProcess` class which will always return `None` and use that check here. Just a way to avoid this `hasattr` magic.

Reply:
Implementing like that and removing `hasattr` and the if block altogether leads to `fromiter` throwing the following error
```
______________________________________________________________________________________________________________________________________________________________
___________________________________________ sympy/stats/tests/test_stochastic_process.py:test_DiscreteMarkovChain ____________________________________________
Traceback (most recent call last):
  File "/home/sidharth/sympy/sympy/stats/tests/test_stochastic_process.py", line 82, in test_DiscreteMarkovChain
    assert Y.joint_distribution(1, Y[2], 3) == JointDistribution(Y[1], Y[2], Y[3])
  File "/home/sidharth/sympy/sympy/stats/stochastic_process_types.py", line 242, in joint_distribution
    expr=Mul.fromiter(arg.pspace.process.density(arg) for arg in args))
  File "/home/sidharth/sympy/sympy/core/basic.py", line 275, in fromiter
    return cls(*tuple(args), **assumptions)
  File "/home/sidharth/sympy/sympy/core/cache.py", line 72, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/home/sidharth/sympy/sympy/core/operations.py", line 53, in __new__
    args = list(map(_sympify_, args))
  File "/home/sidharth/sympy/sympy/core/sympify.py", line 512, in _sympify
    return sympify(a, strict=True)
  File "/home/sidharth/sympy/sympy/core/sympify.py", line 371, in sympify
    raise SympifyError(a)
sympy.core.sympify.SympifyError: SympifyError: None
```

So one way to avoid this is by changing the block to 
```
        density = self.density(args[0])                                         
        if density is None:                                                     
            return JointDistribution(*args)                                     
        density = Lambda(tuple(args),                                           
                expr=Mul.fromiter(arg.pspace.process.density(arg) for arg in args))
        return JointDistributionHandmade(density)
```
Although the tests pass with this, I am not sure if checking what `density` returns on just `args[0]` is a sufficient check. Can you confirm?